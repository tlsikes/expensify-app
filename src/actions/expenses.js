

import uuid from 'uuid';
import database from '../firebase/firebase';

// Actions

// ADD_EXPENSE
export const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
});

export const startAddExpense = (expenseData = {}) => {
    return (dispatch) => {
        const {
            description = '', 
            note = '', 
            amount = 0, 
            createdTimestamp = 0
        } = expenseData;
        const expense = { description, note, amount, createdTimestamp };

        return database.ref('expenses').push(expense).then((ref) => {
            // Dispatch the action for redux
            dispatch(addExpense({
                id: ref.key,
                ...expense
            }));
        });
    }
};

// REMOVE_EXPENSE
export const removeExpense = (
    {
        id = undefined
    } = {}
) => ({
    type: 'REMOVE_EXPENSE',
    id
});

// EDIT_EXPENSE
export const editExpense = (
    id,
    updates
) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
});

export const setExpenses = (expenses) => ({
    type: 'SET_EXPENSES',
    expenses
});

export const startSetExpenses = () => {
    return (dispatch) => {

        return database.ref('expenses')
            .once('value')
            .then((snapshot) => {
                const expenses = [];
                snapshot.forEach((child) => {
                    expenses.push({
                        id: child.key,
                        ...child.val()
                    });
                })
            dispatch(setExpenses(expenses));
        }).catch ((e) => {
            console.log('expense get failed', e);
        });
    }
};

