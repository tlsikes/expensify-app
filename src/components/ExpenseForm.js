import React from 'react';
import { connect } from 'react-redux';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

// momentjs
// AirBNB react-dates

// const now = moment();
// console.log(now.format('MMM Do, YYYY'));

export default class ExpenseForm extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            id: props.expense ? props.expense.id : '',
            description: props.expense ? props.expense.description: '',
            note: props.expense ? props.expense.note: '',
            createdTimestamp: props.expense ? moment(props.expense.createdTimestamp) : moment(),
            amount: props.expense ? (props.expense.amount / 100).toString() : '',
            calendarFocused: false,
            error: ''
        };
    }
    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => ({ description }));
    }
    onAmountChange = (e) => {
        // Note: this does NOT work when passed in as a string, investigate...
        const currencyRegex='/^\d*(\.\d{0,2})?$/';
        const amount = e.target.value;
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({ amount }));
        }
    }
    onDateChange = (createdTimestamp) => {
        if (createdTimestamp) {
            this.setState(() => ({ createdTimestamp }));
        }

    }
    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(() => ({ note }));
    }
    onFocusChange= ({ focused }) => {
        this.setState(() => ({ calendarFocused: focused }));
    }
    onSubmit = (e) => {
        e.preventDefault();
        if (!this.state.description || !this.state.amount) {
            // Set error
            this.setState(() => ({ error: 'Please provide description and amount.' })) ;
        } else {
            this.setState(() => ({ error: '' })) ;
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100,
                createdTimestamp: this.state.createdTimestamp.valueOf(),
                note: this.state.note
            })
        }
    }
    render() {
        return (
            <form className="form" onSubmit={this.onSubmit}>
                { this.state.error && <p className="form__error" color="red">{this.state.error}</p>}
                <input
                    type="text"
                    placeholder="Description"
                    autoFocus
                    className="text-input"
                    value={this.state.description}
                    onChange={this.onDescriptionChange}
                />
                <input
                    type="text"
                    placeholder="Amount"
                    className="text-input"
                    value={this.state.amount}
                    onChange={this.onAmountChange}
                />
                <SingleDatePicker
                    date={this.state.createdTimestamp}
                    onDateChange={this.onDateChange}
                    focused={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    numberOfMonths={1}
                    isOutsideRange={() => false }
                />
                <textarea
                    className="textarea"
                    placeholder="Note for this expense"
                    value={this.state.note}
                    onChange={this.onNoteChange}
                >
                </textarea>
                {/* Wrap button in div it's not a direct child for styling... */}
                <div>
                    <button className="lg-blue-button">Save</button>
                </div>
            </form>
        )
    }
}
