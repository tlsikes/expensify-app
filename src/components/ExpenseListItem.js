import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';

const ExpenseListItem = ({ dispatch, id, description, amount, createdTimestamp }) => (
    <Link className="list-item" to={ `/edit/${id}` }>
        <div>
            <h3 className="list-item__title">{description}</h3>
            <span className="list-item__subtitle">
                {moment(createdTimestamp).format('MMMM Do, YYYY')}
            </span>
        </div>
        <h3 className="list-item__amount">{numeral(amount / 100).format('$0,0.00')}</h3>
    </Link>
);

/*
const mapStateToProps = (state) => {
    return {
        expenses: state.expenses,
        filters: state.filters
    };
}

export default connect(mapStateToProps)(ExpenseList);
*/

export default ExpenseListItem;