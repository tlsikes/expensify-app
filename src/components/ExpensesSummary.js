import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import selectExpenses from '../selectors/expenses';
import totalExpenses from '../selectors/expenses-total';
import numeral from 'numeral';

export const ExpensesSummary = ({ expenseCount, expensesTotal }) => {
    const expenseWord = expenseCount === 1 ? 'expense' : 'expenses';
    const formattedExpenseTotal = numeral(expensesTotal / 100).format('$0,0.00');
    return (
        <div className="page-header">
            <div className="content-container">
                <h2 className="page-header__title">
                    Viewing <span>{expenseCount}</span> {expenseWord} totalling <span>{formattedExpenseTotal}</span>
                </h2>
                <div className="page-header__actions">
                    <Link className="lg-blue-button" to="/create">Add Expense</Link>
                </div>
            </div>
        </div>
        ); 
}

const mapStateToProps = (state) => {
    const selectedExpenses = selectExpenses(state.expenses, state.filters);
    return {
        expenseCount: selectedExpenses.length,
        expensesTotal: totalExpenses(selectedExpenses)
    };
}

export default connect(mapStateToProps)(ExpensesSummary);
