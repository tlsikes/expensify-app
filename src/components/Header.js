import React from 'react';
import { NavLink } from 'react-router-dom';

const Header = () => (
    <header>
        <h1>Expensify</h1>
        <div>
            <NavLink to="/" activeClassName="current-link" exact={true}>Dashboard</NavLink> &nbsp;&nbsp;
            <NavLink to="/create" activeClassName="current-link">New Expense</NavLink> &nbsp;&nbsp;
            <NavLink to="/help" activeClassName="current-link">Help</NavLink> &nbsp;&nbsp;
            <p />
        </div>
    </header>
)

export default Header;
