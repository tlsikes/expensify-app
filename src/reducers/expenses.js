

// Expenses Reducer

const expensesReducerDefaultState = [];

const expensesReducer = (state = expensesReducerDefaultState, action) => {
    switch (action.type) {
        
        case 'ADD_EXPENSE':
            return [
                ...state,
                action.expense
            ];

        case 'EDIT_EXPENSE':
            return state.map((e) => {
                if (e.id === action.id) {
                    return {
                        ...e,
                        ...action.updates
                    }
                } else {
                    return e;
                }
            });

        case 'REMOVE_EXPENSE':
            return state.filter(({ id }) => id != action.id);

        case 'SET_EXPENSES':
            return action.expenses;

        default:
            return state;
    }
};

// Instructor used export default of arrow function directly, not sure how naming works then...
export default expensesReducer;