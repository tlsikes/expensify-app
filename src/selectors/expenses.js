import moment from 'moment';

// Match on text substring..
const isIncluded = (text, substring) => {
    if (typeof text !== 'string') {
        return false;
    }

    if (typeof substring !== 'string') {
        return true;
    }

    if (text.toLowerCase().includes(substring.toLowerCase())) {
        return true;
    }
    return false;
}

// Get visible expenses
const getVisibleExpenses = (expenses, { text, sortBy, startDate, endDate }) => {
    return expenses.filter((e) => {
        const createdAtMoment = moment(e.createdTimestamp);
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;
        const textMatch = isIncluded(e.description, text); // Note description is a string for sure...fix.
        return startDateMatch && endDateMatch && textMatch;
    }).sort((a, b) => {
        if (sortBy === 'amount') {
            return a.amount < b.amount ? 1 : -1;
        } else {
            return a.createdTimestamp < b.createdTimestamp ? 1 : -1;
        }
    });
}

export default getVisibleExpenses;