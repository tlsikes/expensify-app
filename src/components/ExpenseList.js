import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';

export const ExpenseList = (props) => (
    <div className="content-container">
        <div className="list-header">
            <div className="show-for-mobile">Expenses</div>
            <div className="show-for-bigscreen">Expense</div>
            <div className="show-for-bigscreen">Amount</div>
        </div>
        <div className="list-body">
            {
                props.expenses.length === 0 ? (
                    <div className="list-item list-item--message"><h3>No expenses found</h3></div>
                ) : (
                    props.expenses.map((e) => {
                        return (
                            <ExpenseListItem key={e.id} {...e} />
                        );
                    })
                )
            }
        </div>

    </div>
);

const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    };
}

export default connect(mapStateToProps)(ExpenseList);
