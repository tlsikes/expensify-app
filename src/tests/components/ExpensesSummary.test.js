import React from 'react';
import { shallow } from 'enzyme';
import { ExpensesSummary } from '../../components/ExpensesSummary';
import expenses from '../fixtures/expenses';

test('should render with one expense', () => {
    const wrapper = shallow(<ExpensesSummary expenseCount={1} expensesTotal={235} />);
    expect(wrapper).toMatchSnapshot();
})

test('should render with multiple expenses', () => {
    const wrapper = shallow(<ExpensesSummary expenseCount={52} expensesTotal={8289239202} />);
    expect(wrapper).toMatchSnapshot();
})