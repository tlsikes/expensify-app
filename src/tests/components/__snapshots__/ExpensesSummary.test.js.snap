// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render with multiple expenses 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h2
      className="page-header__title"
    >
      Viewing 
      <span>
        52
      </span>
       
      expenses
       totalling 
      <span>
        $82,892,392.02
      </span>
    </h2>
    <div
      className="page-header__actions"
    >
      <Link
        className="lg-blue-button"
        replace={false}
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;

exports[`should render with one expense 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h2
      className="page-header__title"
    >
      Viewing 
      <span>
        1
      </span>
       
      expense
       totalling 
      <span>
        $2.35
      </span>
    </h2>
    <div
      className="page-header__actions"
    >
      <Link
        className="lg-blue-button"
        replace={false}
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;
